Mage/src/main/java/mage/abilities/effects/common/GainsChoiceOfAbilitiesEffect.java

```diff
@@ -0,0 +1,120 @@
+
+package mage.abilities.effects.common;
+
+import mage.abilities.Ability;
+import mage.abilities.Mode;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.continuous.GainAbilityTargetEffect;
+import mage.choices.Choice;
+import mage.choices.ChoiceImpl;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.targetpointer.FixedTargets;
+
+import java.util.*;
+
+/**
+ * Something gains one ability of the player's choice from a list of abilities until end of turn.
+ * Affects any targets set and optionally the source as well. Have no targets to affect only the source.
+ *
+ * @author notgreat, TheElk801
+ */
+public class GainsChoiceOfAbilitiesEffect extends OneShotEffect {
+
+    private final Map<String, Ability> abilityMap = new LinkedHashMap<>();
+    private final boolean affectSource, includeEnd;
+    private final String targetDescription;
+
+    public GainsChoiceOfAbilitiesEffect(Ability... abilities) {
+        this(false, null, true, abilities);
+    }
+
+    public GainsChoiceOfAbilitiesEffect(boolean affectSource, Ability... abilities) {
+        this(affectSource, null, true, abilities);
+    }
+
+    public GainsChoiceOfAbilitiesEffect(boolean affectSource, String targetDescription, boolean includeEnd, Ability... abilities) {
+        super(Outcome.AddAbility);
+        this.affectSource = affectSource;
+        this.targetDescription = targetDescription;
+        this.includeEnd = includeEnd;
+        for (Ability ability : abilities) {
+            this.abilityMap.put(ability.getRule(), ability);
+        }
+    }
+
+    protected GainsChoiceOfAbilitiesEffect(final GainsChoiceOfAbilitiesEffect effect) {
+        super(effect);
+        this.affectSource = effect.affectSource;
+        this.abilityMap.putAll(effect.abilityMap);
+        this.targetDescription = effect.targetDescription;
+        this.includeEnd = effect.includeEnd;
+    }
+
+    @Override
+    public GainsChoiceOfAbilitiesEffect copy() {
+        return new GainsChoiceOfAbilitiesEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        List<Permanent> permanents = new ArrayList<>();
+        if (affectSource) {
+            permanents.add(source.getSourcePermanentIfItStillExists(game));
+        }
+        if (getTargetPointer().isInitialized()) {
```

Comment:
Smells bad. Target already initialized on activate (for one time effects) or from a game start (for continues effects). You don’t need to check it. If it not initialized then report a use case.

![IMG_0640](https://github.com/user-attachments/assets/97ddd7b9-1381-4fd1-817a-275cad08c791)

Reply:
This is a workaround to support having an optional target. Is there a better way to check that? The default target pointer code explicitly checks using `wrongTargetsUsage` which throws an error if no target exists.

I'd prefer to also support having this effect's target pointer be set, but there aren't any cards that need that currently so it's not critical.

I could add a second "check target" boolean I suppose, but that feels just as bad.

Reply:
Ok, looks like you try to make it both: as source effect and as target target. Then it’s better to split it to like AddCountersSourceEffect and AddCountersTargetEffect with some shared code (if it’s possible — but it’s ok to copy paste small parts).

Reply:
How should I share the code like that? Each public class needs to be in a separate file, I could maybe do some weird stuff with inheritance (having a function return what's being affected in one, overriding it in the other) or storing a function or something but those all seem bad.

I could also just not share the code (duplicating it instead), which would be easy but that's a lot of duplicated code. There's also the issue of the one creature which affects both itself and a target.

Reply:
Another alternative -- usage of `TargetController` param like that -- it's used for multi purpose effects:
![shot_240808_182451](https://github.com/user-attachments/assets/8b812713-b172-4a6a-8d3f-e4f57e110369)

Here:
![shot_240808_182541](https://github.com/user-attachments/assets/942afdc8-bcb8-4f4a-9ff0-0c99e1458adf)

Reply:
![shot_240808_182927](https://github.com/user-attachments/assets/9ec7316e-94e1-45a1-9151-d116ce886b35)

Reply:
TargetController seemed to not fit very well (it's all about targeting players/the controller of something), so I made a custom enum inside of the class to handle the three cases.